lib/phoenix_live_view/test/live_view_test.ex

```diff
@@ -376,26 +376,47 @@ defmodule Phoenix.LiveViewTest do
                "some markup in component"
 
   """
-  defmacro render_component(component, assigns, opts \\ []) do
+  defmacro render_component(component, assigns, opts \\ [], do_block \\ []) do
+    {do_block, opts} =
+      case {do_block, opts} do
+        {[do: do_block], _} -> {do_block, opts}
+        {_, [do: do_block]} -> {do_block, []}
+        {_, _} -> {nil, opts}
+      end
+
     endpoint =
       Module.get_attribute(__CALLER__.module, :endpoint) ||
         raise ArgumentError,
               "the module attribute @endpoint is not set for #{inspect(__MODULE__)}"
 
-    quote do
-      Phoenix.LiveViewTest.__render_component__(
-        unquote(endpoint),
-        unquote(component),
-        unquote(assigns),
-        unquote(opts)
-      )
+    socket =
+      quote do
+        %Socket{endpoint: unquote(endpoint), router: unquote(opts)[:router]}
+      end
+
+    if do_block do
+      quote do
+        socket = unquote(socket)
+        var!(assigns) = Map.put(var!(assigns), :socket, socket)
+
+        inner_block = fn _, args ->
+          var!(assigns) = Map.merge(var!(assigns), Map.new(args))
+          unquote(do_block)
+        end
+
+        assigns = unquote(assigns) |> Map.new() |> Map.put(:inner_block, inner_block)
+        Phoenix.LiveViewTest.__render_component__(socket, unquote(component), assigns)
```

Comment:
I see what you mean about the injection of assigns. I think we may need to do this:

```
render_component(socket, assigns, ...) do
  assigns ->
    ...
end
```

So at least we don't have to inject assigns dynamically, what do you think? Then maybe we can do something like this:

```suggestion
        inner_block = fn _, args ->
          unquote({:fn, [], do_block}).(Map.new(args))
        end

        assigns = unquote(assigns) |> Map.new() |> Map.put(:inner_block, inner_block)
        Phoenix.LiveViewTest.__render_component__(socket, unquote(component), assigns)
```

Thoughts?